tensorflow/core/common_runtime/eager/zen_eager_op_rewrite.cc

```diff
@@ -0,0 +1,194 @@
+  }
+
+  // Helper function to insert zen_eager_ops to map.
+  void InsertZenEagerOps(ZenEagerOp op);
+
+  // List of eager ops that can be rewritten with Zen ops. The list is sorted in
+  // alphabetical order.
+  const std::vector<string> kAlwaysRewriteOps = {
+      "AvgPool",          "Conv2D", "FusedBatchNorm", "FusedBatchNormV2",
+      "FusedBatchNormV3", "MatMul", "MaxPool",        "Softmax"};
+};
+
+// The priority value must be higher than MklEagerOpRewrite (10000) so that Zen
+// rewrite happens before Mkl rewrite.
+REGISTER_REWRITE(EagerOpRewriteRegistry::POST_PLACEMENT, 30000,
+                 ZenEagerOpRewrite);
+
+// Constructor
+ZenEagerOpRewrite::ZenEagerOpRewrite(string name, string file, string line)
+    : EagerOpRewrite(name, file, line) {
+  for (const auto &op_name : kAlwaysRewriteOps) {
+    InsertZenEagerOps({op_name, zen_op_registry::GetZenOpName(op_name),
+                       AlwaysRewriteFloat, CreateGenericZenOp});
+  }
+}
+
+void ZenEagerOpRewrite::InsertZenEagerOps(ZenEagerOp op) {
+  zen_eager_ops_.insert(std::make_pair(op.op_name, op));
+}
+
+Status ZenEagerOpRewrite::Run(
+    EagerOperation *orig_op,
+    std::unique_ptr<tensorflow::EagerOperation> *out_op) {
+  if (ShouldRewriteOp(orig_op)) {
+    TF_CHECK_OK(RewriteToZenOp(orig_op, out_op));
+  }
+  return OkStatus();
+}
+
+Status ZenEagerOpRewrite::SetupNewOp(
+    EagerOperation *orig_op, string zen_op_name,
+    std::unique_ptr<EagerOperation> *new_zen_op) {
+  bool is_remote = false;
+  new_zen_op->reset(new tensorflow::EagerOperation(&orig_op->EagerContext()));
+  TF_RETURN_IF_ERROR(new_zen_op->get()->Reset(zen_op_name.c_str(), nullptr,
+                                              is_remote, nullptr));
+
+  int num_inputs = orig_op->Inputs().size();
+  // Add all inputs to the new op.
+  for (int i = 0; i < num_inputs; ++i) {
+    TF_RETURN_IF_ERROR((*new_zen_op)->AddInput(orig_op->Inputs()[i]));
+  }
+
+  // Copy all attributes to the new op.
+  const NodeDef &kOrigNodeDef = orig_op->MutableAttrs()->BuildNodeDef();
+
+  AttrSlice attr_list(kOrigNodeDef);
+  for (const auto &attr : attr_list) {
+    (*new_zen_op)->MutableAttrs()->Set(attr.first, attr.second);
+  }
+
+  (*new_zen_op)->MutableAttrs()->Set("is_eager", true);
+  (*new_zen_op)->MutableAttrs()->Set("reorder_before", false);
+  (*new_zen_op)->MutableAttrs()->Set("reorder_after", false);
+  (*new_zen_op)->MutableAttrs()->Set("in_links", 1);
+  (*new_zen_op)->MutableAttrs()->Set("out_links", 1);
+  (*new_zen_op)->MutableAttrs()->Set("reset", true);
+
+  string device_name = orig_op->DeviceName();
+  return (*new_zen_op)->SetDeviceName(device_name.c_str());
+}
+
+Status ZenEagerOpRewrite::CreateGenericZenOp(
+    EagerOperation *orig_op, std::unique_ptr<EagerOperation> *zen_op,
+    string zen_op_name) {
+  VLOG(1) << " TF-EAGER-REWRITE Info: OriginalOp= " << orig_op->Name()
+          << " ZenOp=" << zen_op_name;
+
+  TF_CHECK_OK(SetupNewOp(orig_op, zen_op_name, zen_op));
```

Comment:
Right now `SetupNewOp` is only used once here. `CreateGenericZenOp` also doesn't do much apart from calling `SetupNewOp`. Do you plan to have another Zen-op creation function that will call `SetupNewOp` in the future? If not, let's just merge `SetupNewOp` and `CreateGenericZenOp` into one function.

Reply:
Accepted the change. Thank you!